### ISSUE TITLE
Zig support

### ISSUE THREAD

**Issue by photex** [fastbuild/fastbuild#983]
Hello!

Zig is interesting in a couple of ways, one of which is that it includes the clang front-end in addition to it's own. `zig cc` and `zig c++` and some others:

```
  ar               Use Zig as a drop-in archiver
  cc               Use Zig as a drop-in C compiler
  c++              Use Zig as a drop-in C++ compiler
  dlltool          Use Zig as a drop-in dlltool.exe
  lib              Use Zig as a drop-in lib.exe
  ranlib           Use Zig as a drop-in ranlib
  objcopy          Use Zig as a drop-in objcopy
```

I thought it might be interesting to test building a mixed zig project using fastbuild (despite 'zig build' existing) since caching and distributed compilation are attractive features of fastbuild.

When looking at the docs (https://fastbuild.org/docs/functions/compiler.html) it seems like I'd have to wrap `zig cc` in a script to start experimenting. 

Would it be useful to perhaps allow some prefix arguments to the compiler? Or perhaps that's just too much of a special case to be worth the trouble?

Additionally, using zig, or even rustc would seem to be straight forward with fastbuild. Has anyone experimented with this yet?

Thanks!
Chip

**Comment by ffulin** [comment]
If Zig can be compiled with a Clang-like compiler (one that supports the same args for preprocessing and various other things as Clang) then it may already work simply be specifying the CompilerFamily in the Compiler(). If that works, the compiler can be added to the compiler detection logic to make it work entirely automatically.

If it doesn't work, because either different args are needed or the behavior differs in some way, as long as equivalent or similar functionality is available, a CompilerDriver can likely be implemented, the comlexity of which will depend on how different the behaviors are.

**Comment by photex** [comment]
Hi! After many many months I'm finding time to resume my experiments.

So, ignoring actual Zig code and only focusing on how to compile C code using `zig cc`, I first tried:

```
// Zig ToolChain
//------------------------------------------------------------------------------
.Compiler           = '$ZigBasePath$\zig.exe'
.CompilerOptions    = 'cc -c %1'      // Input
                    + ' -o"%2"'       // Output
                    + " -D_CRT_SECURE_NO_WARNINGS"

.Librarian          = '$ZigBasePath$\zig.exe'
.LibrarianOptions   = 'lib /out:"%2"'     // Output
                    + ' "%1"'          // Input
                    + ' /WX'           // Warnings as errors
                    + ' /NOLOGO'       // No linker spam

.Linker             = '$ZigBasePath$\zig.exe'
.LinkerOptions      = 'cc -o "%2" "%1"'
```

Fastbuild reports: `FASTBuild Error #1500 - Compiler detection failed. Unrecognized executable`

I tried to create a wrapper script such as `clang.bat`:

```
%~dp0\zig.exe cc %*
```

But this fails, because if you specify the `.bat` extension you get the same error regarding an unrecognized compiler, and if you leave the `.bat` off, then you get another error that the executable can't be found:

```
Error: opening file 'C:\workspaces\slowgames\fastzig\tools\zig\clang' in Compiler ToolManifest
```

Ok, so stepping back and now trying to setup zig using the `Compiler` function:

```
Compiler('zcc')
{
    .Executable = "$ZigBasePath$\zig.exe"
    .CompilerFamily = "clang"
    .ClangRewriteIncludes = false
}

// Zig ToolChain
//------------------------------------------------------------------------------
.Compiler           = 'zcc'
.CompilerOptions    = 'cc -c "%1" -o"%2" -D_CRT_SECURE_NO_WARNINGS'

.Librarian          = '$ZigBasePath$\lib'
.LibrarianOptions   = 'lib /out:"%2"'     // Output
                    + ' "%1"'          // Input
                    + ' /WX'           // Warnings as errors
                    + ' /NOLOGO'       // No linker spam

.Linker             = '$ZigBasePath$\zig.exe'
.LinkerOptions      = 'cc -o "%2" "%1"'
```

But this fails with another error I haven't understood yet:

```
C:\workspaces\slowgames\fastzig>fbuild hello -showcmds
BFF file 'C:\workspaces\slowgames\fastzig\tools\zig.bff' has changed (reparsing will occur).
13>Obj: C:\workspaces\slowgames\fastzig\out\main.obj
C:\workspaces\slowgames\fastzig\tools\zig\zig.exe cc "C:\workspaces\slowgames\fastzig\src\main.c" -D_CRT_SECURE_NO_WARNINGS -E -fdiagnostics-color=always
13>Failed to build Object. Error: 1 (0x01) Target: 'C:\workspaces\slowgames\fastzig\out\main.obj'
13>PROBLEM: C:\workspaces\slowgames\fastzig\out\main.obj
error: AppDataDirUnavailable
FBuild: Error: BUILD FAILED: hello
Time: 0.168s
```

I don't see the `-c` or `-o` flags in the cmd that's printed. I don't know if that is relevant.

Just to make sure I'm not crazy, if I manually use `zig cc` then I can compile the file:

```
tools\zig\zig.exe cc -c src\main.c -o out\main.obj -D_CRT_SECURE_NO_WARNINGS -E -frewrite-includes -fdiagnostics-color=always
tools\zig\zig.exe cc out\main.obj -o out\hello.exe
out\hello.exe

Hello, biff
```

At this point, I'm up against the limits of my knowledge and experience with Fastbuild. I'd appreciate any feedback and help here when you have some spare cycles. :D

**Comment by photex** [comment]
Ok, I peeked around in the Fastbuild source, which clarified that the error is actually coming from `zig cc` here and not from Fastbuild: https://github.com/ziglang/zig/blob/master/lib/std/fs/get_app_data_dir.zig#L10

Zig natively caches build results, so I'm assuming that an expected environment variable isn't present. 

Will investigate.

**Comment by photex** [comment]
Ok, after adding LOCALAPPDATA to the compiler environment there is some slight progress, but still oddly failing:

```
13>PROBLEM: C:\workspaces\slowgames\fastzig\out\glfw\window.obj
error: coff does not support linking multiple objects into one
14>PROBLEM: C:\workspaces\slowgames\fastzig\out\glfw\osmesa_context.obj
error: coff does not support linking multiple objects into one
```

The ball seems to be in Zigs court here, but it's worth mentioning that `zig cc` works with Bazel, CMake/Ninja, it's own `zig build` system, and manually on the command line. Clearly I need to understand the Fastbuild compiler expectations a bit better before I'll make more progress.

**Comment by photex** [comment]
More data:

Ok, testing the same setup on Linux and everything works just fine. This is a Windows specific problem as far as I can tell.

I built my project using a very minimal 'zig.bff' even:

```
.Zig = 'tools/zig/Linux-x86_64/zig'

Compiler('zig')
{
  .Executable = .Zig
  .CompilerFamily = "clang"
}

.Compiler          = 'zig'
.Linker            = .Zig
.CompilerOptions   = ' cc %1 -std=c2x -c -o %2'
.LinkerOptions     = ' cc -fuse-ld=mold %1 -o %2'
```

**Comment by photex** [comment]
So after a bit of noodling, I also tested the zig mingw cross compiler on Linux and it's emitting the same error (even with CMake and when called directly): `error: coff does not support linking multiple objects into one`, So this is a specific limitation of Zig cc for the windows target.

I think we can close this ticket then, because it's clear that in principle `zig cc` will work just fine with Fastbuild as is, and the Windows issue is entirely an issue with Zig and not something triggered by Fastbuild. 

Yay! Fastbuild rules! :D

**Comment by Dandielo** [comment]
Hey, just a small idea you might want to try out. Even though 'zig cc' is a drop in replacement for a compiler, it might not be a drop-in replacement for the linker stage under Windows. However, looking at the "main.zig" file (https://github.com/ziglang/zig/blob/32e88251e48d9f4a412b08acbd04d5694ec91e19/src/main.zig#L298C3-L298C3), there seems to be a "lld-link" command too. You might want to try that one out for your linker part, along with linker specific command line options. 

It might be worth a try if you didn't do it already. I've not used Zig myself, but the problem seems to be interesting.

**Comment by photex** [comment]
Good idea @Dandielo! I gave it a shot, but I get this error before getting to the link phase it seems. Or perhaps the link phase uses the `.Compiler` setting for some reason I'm ignorant of.
